File: apps/luduan-client/src/luduan/translator.py
# Luduan — EPUB to RoboBook narration CLI.
# Copyright (C) 2026  Clementine Pendragon <clem@pendragon.systems>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# Contact: clem@pendragon.systems (Clementine Pendragon, c/o Xian Project Development)

"""Lemonade-backed translation engine for Luduan.

Replaces the original Luduan ``translator.py`` which loaded Qwen
directly onto the GPU.  All inference now goes through Lemonade
Server's ``/v1/chat/completions`` endpoint via the xian-vl pipeline.
"""

import asyncio
import logging

from openai import AsyncOpenAI

from shared_types.constants import DEFAULT_API_URL, DEFAULT_MODEL
from xian.omni_router import OmniModelRouter

logger = logging.getLogger(__name__)


class DocumentTranslator:
    """Translates document passages via Lemonade's chat completions API.

    Parameters
    ----------
    base_url:
        Lemonade API base URL.
    model:
        Model name to use for translation.
    source_lang:
        Source language for the system prompt.
    target_lang:
        Target language for the system prompt.
    router:
        Active model router.
    """

    def __init__(
        self,
        base_url: str = DEFAULT_API_URL,
        model: str = DEFAULT_MODEL,
        source_lang: str = "Chinese",
        target_lang: str = "English",
        router: OmniModelRouter | None = None,
    ) -> None:
        import os
        api_url = os.environ.get("LEMONADE_API_URL", base_url)
        if router:
            api_url = router.api_url
            model = router.llm()
        self._client = AsyncOpenAI(base_url=api_url, api_key="not-needed")
        self._model = model
        self._source = source_lang
        self._target = target_lang

    async def translate(self, text: str) -> str:
        """Translate a single passage.

        Returns the translated text, or the original on failure.
        """
        try:
            response = await self._client.chat.completions.create(
                model=self._model,
                messages=[
                    {
                        "role": "system",
                        "content": (
                            f"You are an expert {self._source}-to-{self._target} translator "
                            f"specialising in web novels (wuxia, xianxia, xuanhuan). "
                            f"Translate the following passage naturally. Preserve proper nouns. "
                            f"Output ONLY the translation."
                        ),
                    },
                    {"role": "user", "content": text},
                ],
                max_tokens=2048,
                extra_body={"chat_template_kwargs": {"enable_thinking": False}},
            )
            return response.choices[0].message.content or text
        except Exception:
            logger.exception("Translation failed for passage (len=%d)", len(text))
            return text

    async def translate_batch(
        self,
        passages: list[str],
        *,
        concurrency: int = 8,
    ) -> list[str]:
        """Translate multiple passages with a bounded concurrency limit."""
        sem = asyncio.Semaphore(max(1, concurrency))

        async def _one(p: str) -> str:
            async with sem:
                return await self.translate(p)

        return await asyncio.gather(*(_one(p) for p in passages))

File: apps/masha-extension/src/core/translator.ts
/*
 * Masha — Browser extension selection translator.
 * Copyright (C) 2026  Clementine Pendragon <clem@pendragon.systems>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 * Contact: clem@pendragon.systems (Clementine Pendragon, c/o Xian Project Development)
 */

/*
 * Browser-free core — see core/constants.ts. The HTTP call is parameterized on
 * a `fetch`-shaped function so this module has no platform dependency at all.
 */

import { normalizeLemonadeBaseUrl } from '../utils/lemonadeUrl';
import { DEFAULT_MAX_TOKENS, DEFAULT_MODEL, DEFAULT_TEMPERATURE } from './constants';
import { buildTranslationMessages, PromptOptions } from './prompt';

/** Mirror of MAGE's reasoning-tag stripping (xian/pipeline.py THINK_*_PAT). */
const THINK_TAGS = /<think>[\s\S]*?<\/think>/gi;
const THINK_OPEN = /<think>[\s\S]*$/i;

/** Characters MAGE trims from the edges of a parsed translation. */
const EDGE_TRIM = /^[\s"'`*•\-]+|[\s"'`*•\-]+$/g;

/** Remove reasoning tags and wrapping punctuation from a model response. */
export function cleanResponse(raw: string): string {
  let cleaned = raw.replace(THINK_TAGS, '');
  cleaned = cleaned.replace(THINK_OPEN, ''); // unmatched/streaming-truncated <think>
  return cleaned.replace(EDGE_TRIM, '').trim();
}

/** Minimal subset of the global `fetch` we depend on (keeps core testable). */
export type FetchFn = (
  input: string,
  init: { method: string; headers: Record<string, string>; body: string },
) => Promise<{ ok: boolean; statusText: string; json(): Promise<any> }>;

export interface TranslateConfig {
  /** Lemonade base URL (with or without ``/v1`` — normalized internally). */
  serverUrl: string;
  model?: string;
  maxTokens?: number;
  temperature?: number;
}

/**
 * Translate a selection via an OpenAI-compatible Lemonade endpoint.
 * Shaped identically to MAGE's chat call so the Python core stays a drop-in.
 */
export async function translate(
  fetchFn: FetchFn,
  config: TranslateConfig,
  opts: PromptOptions,
): Promise<string> {
  const base = normalizeLemonadeBaseUrl(config.serverUrl);
  const endpoint = `${base}/chat/completions`;

  const payload = {
    model: config.model || DEFAULT_MODEL,
    messages: buildTranslationMessages(opts),
    max_tokens: config.maxTokens ?? DEFAULT_MAX_TOKENS,
    temperature: config.temperature ?? DEFAULT_TEMPERATURE,
    stream: false,
  };

  const response = await fetchFn(endpoint, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    throw new Error(`Lemonade API error: ${response.statusText || 'request failed'}`);
  }

  const data = await response.json();
  const content = data?.choices?.[0]?.message?.content;
  const cleaned = cleanResponse(typeof content === 'string' ? content : '');
  if (!cleaned) {
    throw new Error('Empty translation returned');
  }
  return cleaned;
}
